pkg/liqo-controller-manager/external-network/wireguard/wggatewayserver_controller.go

```diff
@@ -0,0 +1,369 @@
+			return ctrl.Result{}, err
+		}
+
+		// Handle status
+		if err := r.handleEndpointStatus(ctx, wgServer, svcNsName, dep, &needStatusUpdate, &requeue); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if err := r.handleSecretRefStatus(ctx, wgServer, &needStatusUpdate, &requeue); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// update status
+		if needStatusUpdate {
+			if err := r.Status().Update(ctx, wgServer); err != nil {
+				klog.Errorf("Unable to update status of %q: %v", req.NamespacedName, err)
+				return ctrl.Result{}, err
+			}
+			klog.Infof("%q status correctly updated", req.NamespacedName)
+		}
+	}
+
+	// Resource is deleting and child resources are deleted as well by garbage collector. Nothing to do.
+	return ctrl.Result{Requeue: requeue}, nil
+}
+
+// SetupWithManager register the WgGatewayServerReconciler to the manager.
+func (r *WgGatewayServerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// TODO: should we watch secrets to reconcile if the secret containing keys change?
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&networkingv1alpha1.WgGatewayServer{}).
+		Owns(&appsv1.Deployment{}).
+		Owns(&corev1.Service{}).
+		Complete(r)
+}
+
+func (r *WgGatewayServerReconciler) ensureDeployment(ctx context.Context, wgServer *networkingv1alpha1.WgGatewayServer,
+	depNsName types.NamespacedName) (*appsv1.Deployment, error) {
+	dep := appsv1.Deployment{ObjectMeta: metav1.ObjectMeta{
+		Name:      depNsName.Name,
+		Namespace: depNsName.Namespace,
+	}}
+
+	op, err := controllerutil.CreateOrUpdate(ctx, r.Client, &dep, func() error {
+		return r.mutateFnWgServerDeployment(&dep, wgServer)
+	})
+	if err != nil {
+		klog.Errorf("error while creating/updating deployment %q (operation: %s): %v", depNsName, op, err)
+		return nil, err
+	}
+
+	klog.Infof("Deployment %q correctly enforced (operation: %s)", depNsName, op)
+	return &dep, nil
+}
+
+func (r *WgGatewayServerReconciler) ensureService(ctx context.Context, wgServer *networkingv1alpha1.WgGatewayServer,
+	svcNsName types.NamespacedName) (*corev1.Service, error) {
+	svc := corev1.Service{ObjectMeta: metav1.ObjectMeta{
+		Name:      svcNsName.Name,
+		Namespace: svcNsName.Namespace,
+	}}
+
+	op, err := controllerutil.CreateOrUpdate(ctx, r.Client, &svc, func() error {
+		return r.mutateFnWgServerService(&svc, wgServer)
+	})
+	if err != nil {
+		klog.Errorf("error while creating/updating service %q (operation: %s): %v", svcNsName, op, err)
+		return nil, err
+	}
+
+	klog.Infof("Service %q correctly enforced (operation: %s)", svcNsName, op)
+	return &svc, nil
+}
+
+func (r *WgGatewayServerReconciler) mutateFnWgServerDeployment(deployment *appsv1.Deployment, wgServer *networkingv1alpha1.WgGatewayServer) error {
+	// Forge metadata
+	deployment.SetLabels(labels.Merge(deployment.GetLabels(), wgServer.Spec.Deployment.Metadata.GetLabels()))
+	deployment.SetAnnotations(labels.Merge(deployment.GetAnnotations(), wgServer.Spec.Deployment.Metadata.GetAnnotations()))
```

Comment:
in this way you will never remove deleted labels and annotations... we have to discuss if this is an expected bahaviour or not